// fp-ts-contrib/lib/Semialign/NonEmptyArray.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @since 0.1.0
 */
var NonEmptyArray_1 = require("fp-ts/lib/NonEmptyArray");
var These_1 = require("fp-ts/lib/These");
var Array_1 = require("../Align/Array");
/**
 * `Semialign` instance for `NonEmptyArray`.
 *
 * @since 0.1.0
 */
exports.semialignNonEmptyArray = {
    URI: NonEmptyArray_1.URI,
    map: NonEmptyArray_1.nonEmptyArray.map,
    /**
     * Apply a function to pairs of elements at the same index in two arrays, collecting the results in a new array.
     * Uses the `These` data type to handle arrays of different lengths.
     *
     * @example
     * import { These } from 'fp-ts/lib/These'
     * import { identity } from 'fp-ts/lib/function'
     * import { alignNonEmptyArray } from 'fp-ts-contrib/lib/Semialign/NonEmptyArray'
     *
     * const f = (x: These<number, string>) => x.fold(a => a.toString(), identity, (a, b) => a + b)
     *
     * assert.deepStrictEqual(semialignNonEmptyArray.alignWith(new NonEmptyArray(1, [2, 3]), new NonEmptyArray('a', ['b', 'c']), f), new NonEmptyArray('1a', ['2b', '3c']))
     * assert.deepStrictEqual(semialignNonEmptyArray.alignWith(new NonEmptyArray(1, [2, 3]), new NonEmptyArray('a', ['b']), f), new NonEmptyArray('1a', ['2b', '3']))
     * assert.deepStrictEqual(semialignNonEmptyArray.alignWith(new NonEmptyArray(1, [2]), new NonEmptyArray('a', ['b', 'c']), f), new NonEmptyArray('1a', ['2b', 'c']))
     *
     * @since 0.1.0
     */
    alignWith: function (fa, fb, f) {
        return NonEmptyArray_1.cons(f(These_1.both(NonEmptyArray_1.head(fa), NonEmptyArray_1.head(fb))), Array_1.alignArray.alignWith(NonEmptyArray_1.tail(fa), NonEmptyArray_1.tail(fb), f));
    },
    /**
     * Takes two arrays and returns an array of corresponding pairs combined using the `These` data type.
     *
     * @example
     * import { These } from 'fp-ts/lib/These'
     * import { identity } from 'fp-ts/lib/function'
     * import { alignNonEmptyArray } from 'fp-ts-contrib/lib/Semialign/NonEmptyArray'
     *
     * assert.deepStrictEqual(semialignNonEmptyArray.align(new NonEmptyArray(1, [2, 3]), new NonEmptyArray('a', ['b', 'c']), new NonEmptyArray(both(1, 'a'), [both(2, 'b'), both(3, 'c')]))
     * assert.deepStrictEqual(semialignNonEmptyArray.align(new NonEmptyArray(1, [2, 3]), new NonEmptyArray('a', ['b']), new NonEmptyArray(both(1, 'a'), [both(2, 'b'), this_(3)])
     * assert.deepStrictEqual(semialignNonEmptyArray.align(new NonEmptyArray(1, [2]), new NonEmptyArray('a', ['b', 'c']), new NonEmptyArray(both(1, 'a'), [both(2, 'b'), that('c')]))
     *
     * @since 0.1.0
     */
    align: function (fa, fb) {
        return NonEmptyArray_1.cons(These_1.both(NonEmptyArray_1.head(fa), NonEmptyArray_1.head(fb)), Array_1.alignArray.align(NonEmptyArray_1.tail(fa), NonEmptyArray_1.tail(fb)));
    }
};
