// @stoplight/ui-kit/CodeViewer/utils/parseCode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const refractor = require("refractor/core");
const lineNumberify_1 = require("./lineNumberify");
refractor.register(require('refractor/lang/bash'));
refractor.register(require('refractor/lang/c'));
refractor.register(require('refractor/lang/csharp'));
refractor.register(require('refractor/lang/diff'));
refractor.register(require('refractor/lang/git'));
refractor.register(require('refractor/lang/go'));
refractor.register(require('refractor/lang/graphql'));
refractor.register(require('refractor/lang/http'));
refractor.register(require('refractor/lang/java'));
refractor.register(require('refractor/lang/javascript'));
refractor.register(require('refractor/lang/json'));
refractor.register(require('refractor/lang/jsx'));
refractor.register(require('refractor/lang/markdown'));
refractor.register(require('refractor/lang/markup'));
refractor.register(require('refractor/lang/markup-templating'));
refractor.register(require('refractor/lang/objectivec'));
refractor.register(require('refractor/lang/ocaml'));
refractor.register(require('refractor/lang/php'));
refractor.register(require('refractor/lang/powershell'));
refractor.register(require('refractor/lang/protobuf'));
refractor.register(require('refractor/lang/python'));
refractor.register(require('refractor/lang/ruby'));
refractor.register(require('refractor/lang/sql'));
refractor.register(require('refractor/lang/swift'));
refractor.register(require('refractor/lang/typescript'));
refractor.register(require('refractor/lang/yaml'));
function parsePlainText(code) {
    return code.split('\n').map((value, i, arr) => ({
        type: 'element',
        tagName: 'span',
        properties: {},
        children: [
            {
                type: 'text',
                value: arr.length - 1 === i ? value : `${value}\n`,
            },
        ],
    }));
}
function safeParse(code, language) {
    if (language) {
        try {
            return refractor.highlight(code, language);
        }
        catch (ex) {
        }
    }
    return parsePlainText(code);
}
function parseCode(code, language, addLineNumbers) {
    try {
        const ast = safeParse(code, language);
        if (addLineNumbers) {
            return lineNumberify_1.lineNumberify(ast);
        }
        return ast;
    }
    catch (ex) {
        return null;
    }
}
exports.parseCode = parseCode;
