// fp-ts-contrib/es6/StateIO.js
/**
 * @since 0.1.0
 */
import * as I from 'fp-ts/es6/IO';
import { getStateM } from 'fp-ts/es6/StateT';
import { pipeable } from 'fp-ts/es6/pipeable';
var T = getStateM(I.io);
/**
 * @since 0.1.0
 */
export var URI = 'StateIO';
/**
 * @since 0.1.0
 */
export function run(ma, s) {
    return ma(s)()[0];
}
/**
 * @since 0.1.0
 */
export var evalState = T.evalState;
/**
 * @since 0.1.0
 */
export var execState = T.execState;
/**
 * @since 0.1.0
 */
export var fromIO = T.fromM;
/**
 * @since 0.1.0
 */
export var fromState = T.fromState;
/**
 * @since 0.1.0
 */
export var get = T.get;
/**
 * @since 0.1.0
 */
export var put = T.put;
/**
 * @since 0.1.0
 */
export var modify = T.modify;
/**
 * @since 0.1.0
 */
export var gets = T.gets;
/**
 * @since 0.1.10
 */
export function fromIOK(f) {
    return function () {
        var a = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            a[_i] = arguments[_i];
        }
        return fromIO(f.apply(void 0, a));
    };
}
/**
 * @since 0.1.10
 */
export function chainIOK(f) {
    return chain(fromIOK(f));
}
/**
 * @since 0.1.0
 */
export var stateIO = {
    URI: URI,
    map: T.map,
    of: T.of,
    ap: T.ap,
    chain: T.chain
};
var _a = pipeable(stateIO), ap = _a.ap, apFirst = _a.apFirst, apSecond = _a.apSecond, chain = _a.chain, chainFirst = _a.chainFirst, flatten = _a.flatten, map = _a.map;
export { 
/**
 * @since 0.1.0
 */
ap, 
/**
 * @since 0.1.0
 */
apFirst, 
/**
 * @since 0.1.0
 */
apSecond, 
/**
 * @since 0.1.0
 */
chain, 
/**
 * @since 0.1.0
 */
chainFirst, 
/**
 * @since 0.1.0
 */
flatten, 
/**
 * @since 0.1.0
 */
map };
