// fp-ts-contrib/es6/StateEither.js
/**
 * @since 0.1.12
 */
import * as E from 'fp-ts/es6/Either';
import { getStateM } from 'fp-ts/es6/StateT';
import { pipeable } from 'fp-ts/es6/pipeable';
var T = getStateM(E.either);
/**
 * @since 0.1.12
 */
export var URI = 'StateEither';
/**
 * @since 0.1.12
 */
export var evalState = T.evalState;
/**
 * @since 0.1.12
 */
export var execState = T.execState;
/**
 * @since 0.1.12
 */
export function left(e) {
    return fromEither(E.left(e));
}
/**
 * @since 0.1.12
 */
export var right = T.of;
/**
 * @since 0.1.12
 */
export var rightState = T.fromState;
/**
 * @since 0.1.12
 */
export function leftState(me) {
    return function (s) { return E.left(me(s)[0]); };
}
/**
 * @since 0.1.12
 */
export var get = T.get;
/**
 * @since 0.1.12
 */
export var put = T.put;
/**
 * @since 0.1.12
 */
export var modify = T.modify;
/**
 * @since 0.1.12
 */
export var gets = T.gets;
/**
 * @since 0.1.0
 */
export var fromEither = T.fromM;
/**
 * @since 0.1.12
 */
export function fromEitherK(f) {
    return function () {
        var a = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            a[_i] = arguments[_i];
        }
        return fromEither(f.apply(void 0, a));
    };
}
/**
 * @since 0.1.12
 */
export function chainEitherK(f) {
    return chain(fromEitherK(f));
}
/**
 * @since 0.1.12
 */
export var stateEither = {
    URI: URI,
    map: T.map,
    of: right,
    ap: T.ap,
    chain: T.chain,
    throwError: left
};
var _a = pipeable(stateEither), ap = _a.ap, apFirst = _a.apFirst, apSecond = _a.apSecond, chain = _a.chain, chainFirst = _a.chainFirst, flatten = _a.flatten, map = _a.map, filterOrElse = _a.filterOrElse, fromOption = _a.fromOption, fromPredicate = _a.fromPredicate;
export { 
/**
 * @since 0.1.12
 */
ap, 
/**
 * @since 0.1.12
 */
apFirst, 
/**
 * @since 0.1.12
 */
apSecond, 
/**
 * @since 0.1.12
 */
chain, 
/**
 * @since 0.1.12
 */
chainFirst, 
/**
 * @since 0.1.12
 */
flatten, 
/**
 * @since 0.1.12
 */
map, 
/**
 * @since 0.1.12
 */
filterOrElse, 
/**
 * @since 0.1.12
 */
fromOption, 
/**
 * @since 0.1.12
 */
fromPredicate };
