// fp-ts-contrib/es6/StateTaskEither.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { getStateM } from 'fp-ts/es6/StateT';
import * as TE from 'fp-ts/es6/TaskEither';
import { pipeable } from 'fp-ts/es6/pipeable';
var T = getStateM(TE.taskEither);
/**
 * @since 0.1.0
 */
export var URI = 'StateTaskEither';
/**
 * @since 0.1.0
 */
export function run(ma, s) {
    return ma(s)();
}
/**
 * @since 0.1.0
 */
export var evalState = T.evalState;
/**
 * @since 0.1.0
 */
export var execState = T.execState;
/**
 * @since 0.1.0
 */
export function left(e) {
    return fromTaskEither(TE.left(e));
}
/**
 * @since 0.1.0
 */
export var right = T.of;
/**
 * @since 0.1.0
 */
export function rightTask(ma) {
    return fromTaskEither(TE.rightTask(ma));
}
/**
 * @since 0.1.0
 */
export function leftTask(me) {
    return fromTaskEither(TE.leftTask(me));
}
/**
 * @since 0.1.0
 */
export var fromTaskEither = T.fromM;
/**
 * @since 0.1.0
 */
export function fromIOEither(ma) {
    return fromTaskEither(TE.fromIOEither(ma));
}
/**
 * @since 0.1.0
 */
export function rightIO(ma) {
    return fromTaskEither(TE.rightIO(ma));
}
/**
 * @since 0.1.0
 */
export function leftIO(me) {
    return fromTaskEither(TE.leftIO(me));
}
/**
 * @since 0.1.0
 */
export var rightState = T.fromState;
/**
 * @since 0.1.0
 */
export function leftState(me) {
    return function (s) { return TE.left(me(s)[0]); };
}
/**
 * @since 0.1.0
 */
export var get = T.get;
/**
 * @since 0.1.0
 */
export var put = T.put;
/**
 * @since 0.1.0
 */
export var modify = T.modify;
/**
 * @since 0.1.0
 */
export var gets = T.gets;
/**
 * @since 0.1.10
 */
export function fromEitherK(f) {
    return function () {
        var a = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            a[_i] = arguments[_i];
        }
        return fromEither(f.apply(void 0, a));
    };
}
/**
 * @since 0.1.10
 */
export function chainEitherK(f) {
    return chain(fromEitherK(f));
}
/**
 * @since 0.1.10
 */
export function fromIOEitherK(f) {
    return function () {
        var a = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            a[_i] = arguments[_i];
        }
        return fromIOEither(f.apply(void 0, a));
    };
}
/**
 * @since 0.1.10
 */
export function chainIOEitherK(f) {
    return chain(fromIOEitherK(f));
}
/**
 * @since 0.1.10
 */
export function fromTaskEitherK(f) {
    return function () {
        var a = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            a[_i] = arguments[_i];
        }
        return fromTaskEither(f.apply(void 0, a));
    };
}
/**
 * @since 0.1.10
 */
export function chainTaskEitherK(f) {
    return chain(fromTaskEitherK(f));
}
/**
 * @since 0.1.0
 */
export var stateTaskEither = {
    URI: URI,
    map: T.map,
    of: right,
    ap: T.ap,
    chain: T.chain,
    throwError: left
};
/**
 * Like `stateTaskEither` but `ap` is sequential
 * @since 0.1.0
 */
export var stateTaskEitherSeq = __assign(__assign({}, stateTaskEither), { ap: function (mab, ma) { return T.chain(mab, function (f) { return T.map(ma, f); }); } });
var _a = pipeable(stateTaskEither), ap = _a.ap, apFirst = _a.apFirst, apSecond = _a.apSecond, chain = _a.chain, chainFirst = _a.chainFirst, flatten = _a.flatten, map = _a.map, filterOrElse = _a.filterOrElse, fromEither = _a.fromEither, fromOption = _a.fromOption, fromPredicate = _a.fromPredicate;
export { 
/**
 * @since 0.1.0
 */
ap, 
/**
 * @since 0.1.0
 */
apFirst, 
/**
 * @since 0.1.0
 */
apSecond, 
/**
 * @since 0.1.0
 */
chain, 
/**
 * @since 0.1.0
 */
chainFirst, 
/**
 * @since 0.1.0
 */
flatten, 
/**
 * @since 0.1.0
 */
map, 
/**
 * @since 0.1.0
 */
filterOrElse, 
/**
 * @since 0.1.0
 */
fromEither, 
/**
 * @since 0.1.0
 */
fromOption, 
/**
 * @since 0.1.0
 */
fromPredicate };
