// @stoplight/elements/hooks/useProjectNodes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const lodash_1 = require("lodash");
const React = tslib_1.__importStar(require("react"));
const Provider_1 = require("../containers/Provider");
const srns_1 = require("../utils/srns");
const useRequest_1 = require("./useRequest");
const MAX_PAGE_SIZE = 300;
function useProjectNodes(srn, opts = {}) {
    const projectSrn = srns_1.serializeSrn(Object.assign(Object.assign({}, srns_1.deserializeSrn(srn)), { uri: undefined }));
    const projectToken = React.useContext(Provider_1.ProjectTokenContext);
    const { isLoading, data, error } = useRequest_1.useRequest({
        pathname: '/projects.nodes',
        params: {
            srn: projectSrn,
            group: opts.group,
            first: MAX_PAGE_SIZE,
            token: projectToken,
        },
    });
    return {
        data: Object.assign(Object.assign({}, data), { items: data ? lodash_1.uniqBy(data.items, 'srn') : [] }),
        isLoading,
        error,
    };
}
exports.useProjectNodes = useProjectNodes;
