// @stoplight/http-spec/oas2/transformers/responses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = require("lodash");
const params_1 = require("./params");
function translateToResponse(produces, response, statusCode) {
    const headers = params_1.translateToHeaderParams(response.headers || {});
    const objectifiedExamples = lodash_1.chain(response.examples)
        .mapValues((value, key) => ({ key, value }))
        .values()
        .value();
    const contents = produces.map(produceElement => ({
        mediaType: produceElement,
        schema: response.schema,
        examples: objectifiedExamples.filter(example => example.key === produceElement),
    }));
    const translatedResponses = {
        code: statusCode,
        description: response.description,
        headers,
        contents,
    };
    const foreignExamples = objectifiedExamples.filter(example => !produces.includes(example.key));
    if (foreignExamples.length > 0) {
        if (translatedResponses.contents.length === 0)
            translatedResponses.contents[0] = {
                mediaType: '',
                schema: {},
                examples: [],
            };
        translatedResponses.contents[0].examples.push(...foreignExamples);
    }
    return translatedResponses;
}
function translateToResponses(responses, produces) {
    return lodash_1.map(responses, lodash_1.partial(translateToResponse, produces));
}
exports.translateToResponses = translateToResponses;
