// @stoplight/scripts/dist/commands/release/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const command_1 = require("@oclif/command");
const utils_1 = require("../../utils");
class ReleaseCommand extends command_1.Command {
    run() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const parsed = this.parse(ReleaseCommand);
            const commands = [];
            commands.push(utils_1.buildCommand('semantic-release', {
                defaultArgs: {},
                rawArgs: parsed.raw,
                flags: Object.keys(ReleaseCommand.flags),
            }));
            if (parsed.flags.verbose) {
                this.log(`commands:`);
                for (const command of commands) {
                    this.log(`    '${command}'`);
                }
            }
            for (const command of commands) {
                utils_1.runCommand(command);
            }
        });
    }
}
exports.default = ReleaseCommand;
ReleaseCommand.strict = false;
ReleaseCommand.description = 'Publish new src or docs release.';
ReleaseCommand.examples = [`$ sl-scripts release`, `$ sl-scripts release:docs`];
ReleaseCommand.args = [];
ReleaseCommand.flags = {
    verbose: command_1.flags.boolean({
        description: 'moar logs',
        required: false,
    }),
};
