// fp-ts-contrib/lib/ArrayOption.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Array_1 = require("fp-ts/lib/Array");
var OptionT_1 = require("fp-ts/lib/OptionT");
var pipeable_1 = require("fp-ts/lib/pipeable");
var T = OptionT_1.getOptionM(Array_1.array);
/**
 * @since 0.1.0
 */
exports.URI = 'ArrayOption';
/**
 * @since 0.1.0
 */
exports.fromArray = T.fromM;
/**
 * @since 0.1.0
 */
exports.fromOption = Array_1.of;
/**
 * @since 0.1.0
 */
exports.none = T.none();
/**
 * @since 0.1.0
 */
exports.some = T.of;
/**
 * @since 0.1.0
 */
function fold(onNone, onSome) {
    return function (as) { return T.fold(as, onNone, onSome); };
}
exports.fold = fold;
/**
 * @since 0.1.0
 */
function getOrElse(onNone) {
    return function (as) { return T.getOrElse(as, onNone); };
}
exports.getOrElse = getOrElse;
/**
 * @since 0.1.10
 */
function fromOptionK(f) {
    return function () {
        var a = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            a[_i] = arguments[_i];
        }
        return exports.fromOption(f.apply(void 0, a));
    };
}
exports.fromOptionK = fromOptionK;
/**
 * @since 0.1.10
 */
function chainOptionK(f) {
    return chain(fromOptionK(f));
}
exports.chainOptionK = chainOptionK;
/**
 * @since 0.1.0
 */
exports.arrayOption = {
    URI: exports.URI,
    map: T.map,
    of: exports.some,
    ap: T.ap,
    chain: T.chain,
    alt: T.alt
};
var _a = pipeable_1.pipeable(exports.arrayOption), alt = _a.alt, ap = _a.ap, apFirst = _a.apFirst, apSecond = _a.apSecond, chain = _a.chain, chainFirst = _a.chainFirst, flatten = _a.flatten, map = _a.map;
exports.alt = alt;
exports.ap = ap;
exports.apFirst = apFirst;
exports.apSecond = apSecond;
exports.chain = chain;
exports.chainFirst = chainFirst;
exports.flatten = flatten;
exports.map = map;
