// fp-ts-contrib/lib/IOOption.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var O = require("fp-ts/lib/Option");
var OptionT_1 = require("fp-ts/lib/OptionT");
var Filterable_1 = require("fp-ts/lib/Filterable");
var IO_1 = require("fp-ts/lib/IO");
var pipeable_1 = require("fp-ts/lib/pipeable");
var T = OptionT_1.getOptionM(IO_1.io);
/**
 * @since 0.1.14
 */
exports.URI = 'IOOption';
/**
 * @since 0.1.14
 */
exports.none = T.none();
/**
 * @since 0.1.14
 */
exports.some = T.of;
/**
 * @since 0.1.14
 */
exports.fromIO = T.fromM;
/**
 * @since 0.1.14
 */
exports.fromOption = IO_1.of;
/**
 * @since 0.1.14
 */
function fold(onNone, onSome) {
    return function (ma) { return T.fold(ma, onNone, onSome); };
}
exports.fold = fold;
/**
 * @since 0.1.14
 */
function fromNullable(a) {
    return exports.fromOption(O.fromNullable(a));
}
exports.fromNullable = fromNullable;
/**
 * @since 0.1.14
 */
function getOrElse(onNone) {
    return function (ma) { return T.getOrElse(ma, onNone); };
}
exports.getOrElse = getOrElse;
/**
 * @since 0.1.14
 */
function fromIOEither(ma) {
    return IO_1.io.map(ma, O.fromEither);
}
exports.fromIOEither = fromIOEither;
/**
 * @since 0.1.14
 */
function toUndefined(ma) {
    return IO_1.io.map(ma, O.toUndefined);
}
exports.toUndefined = toUndefined;
/**
 * @since 0.1.14
 */
function toNullable(ma) {
    return IO_1.io.map(ma, O.toNullable);
}
exports.toNullable = toNullable;
/**
 * @since 0.1.14
 */
function mapNullable(f) {
    return IO_1.map(O.mapNullable(f));
}
exports.mapNullable = mapNullable;
/**
 * @since 0.1.14
 */
function getApplySemigroup(S) {
    return IO_1.getSemigroup(O.getApplySemigroup(S));
}
exports.getApplySemigroup = getApplySemigroup;
/**
 * @since 0.1.14
 */
function getApplyMonoid(M) {
    return {
        concat: getApplySemigroup(M).concat,
        empty: exports.some(M.empty)
    };
}
exports.getApplyMonoid = getApplyMonoid;
/**
 * @since 0.1.14
 */
function fromOptionK(f) {
    return function () {
        var a = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            a[_i] = arguments[_i];
        }
        return exports.fromOption(f.apply(void 0, a));
    };
}
exports.fromOptionK = fromOptionK;
/**
 * @since 0.1.14
 */
function chainOptionK(f) {
    return chain(fromOptionK(f));
}
exports.chainOptionK = chainOptionK;
/**
 * @since 0.1.14
 */
exports.ioOption = __assign({ URI: exports.URI, map: T.map, of: exports.some, ap: T.ap, chain: T.chain, alt: T.alt, fromIO: exports.fromIO }, Filterable_1.getFilterableComposition(IO_1.io, O.option));
var _a = pipeable_1.pipeable(exports.ioOption), alt = _a.alt, ap = _a.ap, apFirst = _a.apFirst, apSecond = _a.apSecond, chain = _a.chain, chainFirst = _a.chainFirst, flatten = _a.flatten, map = _a.map, partition = _a.partition, partitionMap = _a.partitionMap, filter = _a.filter, filterMap = _a.filterMap, compact = _a.compact, separate = _a.separate;
exports.alt = alt;
exports.ap = ap;
exports.apFirst = apFirst;
exports.apSecond = apSecond;
exports.chain = chain;
exports.chainFirst = chainFirst;
exports.flatten = flatten;
exports.map = map;
exports.partition = partition;
exports.partitionMap = partitionMap;
exports.filter = filter;
exports.filterMap = filterMap;
exports.compact = compact;
exports.separate = separate;
