// fp-ts-contrib/lib/ReaderIO.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @since 0.1.0
 */
var I = require("fp-ts/lib/IO");
var pipeable_1 = require("fp-ts/lib/pipeable");
var ReaderT_1 = require("fp-ts/lib/ReaderT");
var T = ReaderT_1.getReaderM(I.io);
/**
 * @since 0.1.0
 */
exports.URI = 'ReaderIO';
/**
 * @since 0.1.0
 */
function run(ma, r) {
    return ma(r)();
}
exports.run = run;
/**
 * @since 0.1.0
 */
exports.fromReader = T.fromReader;
/**
 * @since 0.1.0
 */
exports.fromIO = T.fromM;
/**
 * @since 0.1.0
 */
exports.ask = T.ask;
/**
 * @since 0.1.0
 */
exports.asks = T.asks;
/**
 * @since 0.1.0
 */
function local(f) {
    return function (ma) { return T.local(ma, f); };
}
exports.local = local;
/**
 * @since 0.1.10
 */
function fromIOK(f) {
    return function () {
        var a = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            a[_i] = arguments[_i];
        }
        return exports.fromIO(f.apply(void 0, a));
    };
}
exports.fromIOK = fromIOK;
/**
 * @since 0.1.10
 */
function chainIOK(f) {
    return chain(fromIOK(f));
}
exports.chainIOK = chainIOK;
/**
 * @since 0.1.0
 */
exports.readerIO = {
    URI: exports.URI,
    map: T.map,
    of: T.of,
    ap: T.ap,
    chain: T.chain
};
var _a = pipeable_1.pipeable(exports.readerIO), ap = _a.ap, apFirst = _a.apFirst, apSecond = _a.apSecond, chain = _a.chain, chainFirst = _a.chainFirst, flatten = _a.flatten, map = _a.map;
exports.ap = ap;
exports.apFirst = apFirst;
exports.apSecond = apSecond;
exports.chain = chain;
exports.chainFirst = chainFirst;
exports.flatten = flatten;
exports.map = map;
