// fp-ts-contrib/lib/StateIO.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @since 0.1.0
 */
var I = require("fp-ts/lib/IO");
var StateT_1 = require("fp-ts/lib/StateT");
var pipeable_1 = require("fp-ts/lib/pipeable");
var T = StateT_1.getStateM(I.io);
/**
 * @since 0.1.0
 */
exports.URI = 'StateIO';
/**
 * @since 0.1.0
 */
function run(ma, s) {
    return ma(s)()[0];
}
exports.run = run;
/**
 * @since 0.1.0
 */
exports.evalState = T.evalState;
/**
 * @since 0.1.0
 */
exports.execState = T.execState;
/**
 * @since 0.1.0
 */
exports.fromIO = T.fromM;
/**
 * @since 0.1.0
 */
exports.fromState = T.fromState;
/**
 * @since 0.1.0
 */
exports.get = T.get;
/**
 * @since 0.1.0
 */
exports.put = T.put;
/**
 * @since 0.1.0
 */
exports.modify = T.modify;
/**
 * @since 0.1.0
 */
exports.gets = T.gets;
/**
 * @since 0.1.10
 */
function fromIOK(f) {
    return function () {
        var a = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            a[_i] = arguments[_i];
        }
        return exports.fromIO(f.apply(void 0, a));
    };
}
exports.fromIOK = fromIOK;
/**
 * @since 0.1.10
 */
function chainIOK(f) {
    return chain(fromIOK(f));
}
exports.chainIOK = chainIOK;
/**
 * @since 0.1.0
 */
exports.stateIO = {
    URI: exports.URI,
    map: T.map,
    of: T.of,
    ap: T.ap,
    chain: T.chain
};
var _a = pipeable_1.pipeable(exports.stateIO), ap = _a.ap, apFirst = _a.apFirst, apSecond = _a.apSecond, chain = _a.chain, chainFirst = _a.chainFirst, flatten = _a.flatten, map = _a.map;
exports.ap = ap;
exports.apFirst = apFirst;
exports.apSecond = apSecond;
exports.chain = chain;
exports.chainFirst = chainFirst;
exports.flatten = flatten;
exports.map = map;
