// @stoplight/elements/components/Search/List.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = require("@stoplight/path");
const ui_kit_1 = require("@stoplight/ui-kit");
const ScrollContainer_1 = require("@stoplight/ui-kit/ScrollContainer");
const classnames_1 = tslib_1.__importDefault(require("classnames"));
const React = tslib_1.__importStar(require("react"));
const useComponents_1 = require("../../hooks/useComponents");
const node_1 = require("../../utils/node");
exports.NodeList = ({ isLoading, error, nodes, onReset, onClose }) => {
    if (error) {
        return (React.createElement(ui_kit_1.NonIdealState, { title: "An error has occured!", description: "Try refreshing the page. If the error persists, please reach out to us at support@stoplight.io.", icon: "error", action: React.createElement(ui_kit_1.Button, { text: "Reload the Page", onClick: () => {
                    window.location.reload();
                } }) }));
    }
    if (!nodes || !nodes.length) {
        if (!nodes && isLoading) {
            return React.createElement(ui_kit_1.Spinner, { className: "mt-32" });
        }
        else {
            return (React.createElement(ui_kit_1.NonIdealState, { title: "No Results", description: "Try tweaking your filters or search term.", icon: "zoom-out", action: React.createElement(ui_kit_1.Button, { text: "Clear Search & Filters", onClick: onReset }) }));
        }
    }
    return (React.createElement(ScrollContainer_1.ScrollContainer, { className: "NodeList" },
        nodes.map((item, i) => (React.createElement(NodeListItem, { key: i, item: item, isLoading: isLoading, onClose: onClose, onReset: onReset }))),
        isLoading && (React.createElement("div", { className: "mt-2 mb-8" },
            React.createElement(ui_kit_1.Spinner, { className: "mt-2" })))));
};
const NodeListItem = ({ isLoading, item, onReset, onClose }) => {
    const components = useComponents_1.useComponents();
    const { orgSlug, projectSlug } = path_1.deserializeSrn(item.srn);
    const onClick = React.useCallback(e => {
        if (onReset) {
            onReset(e);
        }
        if (onClose) {
            onClose();
        }
    }, [onClose, onReset]);
    let dataContext = null;
    if (item.data && item.data.match('<em>')) {
        dataContext = (React.createElement(ui_kit_1.Callout, { style: { maxHeight: 150 }, className: classnames_1.default('mt-4 -mb-1 -mx-1 overflow-auto', {
                [ui_kit_1.Classes.SKELETON]: isLoading,
            }) },
            React.createElement(HighlightSearchContext, { markup: item.data })));
    }
    const children = (React.createElement("div", { key: "1", className: "NodeList__item flex px-6 py-8 border-b cursor-pointer dark:border-lighten-4 hover:bg-gray-1 dark-hover:bg-lighten-3", onClick: onClick },
        React.createElement("div", { className: "mr-4" },
            React.createElement(ui_kit_1.Tag, { icon: node_1.NodeTypeIcons[item.type] && React.createElement(ui_kit_1.Icon, { icon: node_1.NodeTypeIcons[item.type], iconSize: 11 }), style: { backgroundColor: node_1.NodeTypeColors[item.type] || undefined }, title: node_1.NodeTypePrettyName[item.type] || item.type, className: "py-1 dark:text-white" })),
        React.createElement("div", { className: "flex-1" },
            React.createElement("div", { className: "flex items-center" },
                React.createElement("div", { className: classnames_1.default(ui_kit_1.Classes.HEADING, 'inline-block flex items-center m-0', {
                        [ui_kit_1.Classes.SKELETON]: isLoading,
                    }) },
                    React.createElement(HighlightSearchContext, { markup: item.name || 'No Name...' })),
                React.createElement("div", { className: "flex-1" }),
                React.createElement("div", { className: classnames_1.default(ui_kit_1.Classes.TEXT_MUTED, 'flex text-sm', {
                        [ui_kit_1.Classes.SKELETON]: isLoading,
                    }) },
                    React.createElement("div", null, orgSlug),
                    React.createElement("div", { className: "px-1" }, "/"),
                    React.createElement("div", null, projectSlug))),
            item.summary && (React.createElement("div", { className: "flex" },
                React.createElement("div", { className: classnames_1.default('flex-1 mt-2', {
                        [ui_kit_1.Classes.SKELETON]: isLoading,
                    }) },
                    React.createElement(HighlightSearchContext, { markup: item.summary })))),
            dataContext)));
    if (components.link) {
        return components.link({
            node: {
                className: 'reset',
                url: item.srn,
            },
            children,
            defaultComponents: components,
            parent: {},
            path: [],
        }, item.id);
    }
    return children;
};
const HighlightSearchContext = ({ markup, className }) => {
    return (React.createElement("div", { className: classnames_1.default('Search__highlight whitespace-pre-wrap', className), dangerouslySetInnerHTML: { __html: markup } }));
};
