// @stoplight/elements/components/TableOfContents/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const TableOfContents_1 = require("@stoplight/ui-kit/TableOfContents");
const React = tslib_1.__importStar(require("react"));
const useComponents_1 = require("../../hooks/useComponents");
const useComputeToc_1 = require("../../hooks/useComputeToc");
exports.TableOfContents = (_a) => {
    var { contents: _contents, items = [], srn } = _a, props = tslib_1.__rest(_a, ["contents", "items", "srn"]);
    const hasContents = _contents && _contents.length;
    let contents = useComputeToc_1.useComputeToc(hasContents ? [] : items);
    if (_contents && hasContents) {
        contents = _contents;
    }
    const components = useComponents_1.useComponents();
    const rowRenderer = React.useCallback((_item, DefaultRow) => {
        const item = Object.assign(Object.assign({}, _item), { isActive: _item.href ? _item.href === srn : false });
        if (components.link && item.href) {
            return components.link({
                node: {
                    title: item.name,
                    url: item.href,
                    className: 'reset',
                },
                children: React.createElement(DefaultRow, { key: item.href, item: item }),
            }, item.href);
        }
        return React.createElement(DefaultRow, { key: item.href || item.id, item: item });
    }, [srn, components]);
    return React.createElement(TableOfContents_1.TableOfContents, Object.assign({}, props, { contents: contents, rowRenderer: rowRenderer }));
};
