// @stoplight/http-spec/postman/transformers/params.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("@stoplight/types");
const jsonSchemaGenerator = require("json-schema-generator");
const typeIs = require("type-is");
const util_1 = require("../util");
function transformQueryParam(queryParam) {
    return Object.assign({ name: queryParam.key || '', style: types_1.HttpParamStyles.Form, required: true }, (queryParam.value ? util_1.transformStringValueToSchema(queryParam.value) : undefined));
}
exports.transformQueryParam = transformQueryParam;
function transformHeader(header) {
    return Object.assign({ name: header.key.toLowerCase(), style: types_1.HttpParamStyles.Simple, required: true }, (header.value ? util_1.transformStringValueToSchema(header.value) : undefined));
}
exports.transformHeader = transformHeader;
function transformPathParams(segments) {
    return segments.reduce((params, segment) => {
        if (segment.startsWith(':')) {
            params.push({
                name: segment.substring(1),
                style: types_1.HttpParamStyles.Simple,
                required: true,
            });
        }
        return params;
    }, []);
}
exports.transformPathParams = transformPathParams;
function transformBody(body, mediaType) {
    switch (body.mode) {
        case 'raw':
            if (!body.raw)
                return;
            return { contents: [transformRawBody(body.raw, mediaType)] };
        case 'formdata':
            if (!body.formdata)
                return;
            return {
                contents: [transformParamsBody(body.formdata, mediaType || 'multipart/form-data')],
            };
        case 'urlencoded':
            if (!body.urlencoded)
                return;
            return {
                contents: [transformParamsBody(body.urlencoded, mediaType || 'application/x-www-form-urlencoded')],
            };
    }
    return;
}
exports.transformBody = transformBody;
function transformRawBody(raw, mediaType = 'text/plain') {
    if (typeIs.is(mediaType, ['application/json', 'application/*+json'])) {
        try {
            const parsed = JSON.parse(raw);
            return {
                mediaType,
                examples: [
                    {
                        key: 'default',
                        value: parsed,
                    },
                ],
                schema: jsonSchemaGenerator(parsed),
            };
        }
        catch (e) {
        }
    }
    return {
        mediaType,
        examples: [
            {
                key: 'default',
                value: raw,
            },
        ],
    };
}
exports.transformRawBody = transformRawBody;
function transformParamsBody(params, mediaType) {
    const paramsList = params.map(item => {
        return {
            name: item.key || generateId(),
            schema: {
                type: 'string',
                description: item.description && util_1.transformDescriptionDefinition(item.description),
            },
            value: item.value,
        };
    });
    return {
        mediaType,
        schema: {
            type: 'object',
            properties: paramsList.reduce((props, param) => {
                props[param.name] = param.schema;
                return props;
            }, {}),
        },
        examples: [
            {
                key: 'default',
                value: paramsList.reduce((values, param) => {
                    values[param.name] = param.value;
                    return values;
                }, {}),
            },
        ],
    };
}
function generateId() {
    return ('_gen_' +
        Math.round(Math.pow(8, 6) * Math.random())
            .toString(16)
            .padStart(6, '0'));
}
